framework/security/src/main/java/org/apache/cloudstack/framework/security/keystore/KeystoreManagerImpl.java

```diff
@@ -109,15 +109,15 @@ public byte[] getKeystoreBits(String name, String aliasForCertificateInStore, St
         try {
             return CertificateHelper.buildAndSaveKeystore(certs, storePassword);
         } catch (KeyStoreException e) {
-            s_logger.warn("Unable to build keystore for " + name + " due to KeyStoreException");
+            s_logger.warn("Unable to build keystore for " + name + " due to KeyStoreException", e);
```

Comment:
Doing like this,
```suggestion
            s_logger.warn(String.format("Unable to build keystore for %s due to %s", name, e.getClass().getSimpleName()));
            if (s_logger.isDebugEnabled()) {
                s_logger.debug(String.format("Unable to build keystore for %s due to %s", name, e.getClass().getSimpleName()), e);
            }
```
the catch clauses can be unified to one
```
            } catch (KeyStoreException | CertificateException | InvalidKeySpecException | IOException e) {
```

Reply:
Thanks, done

Reply:
@DaanHoogland 
I have a question, is the check `if (s_logger.isDebugEnabled()) {` necessary ?

Reply:
not if the string is generated separately anyway. I think in this case we could leave it out, @weizhouapache

Reply:
OK. thanks @DaanHoogland